src/ducc0/infra/threading.cc

```diff
@@ -367,29 +369,73 @@ class thread_pool
       }
   };
 
-inline thread_pool &get_pool()
+// return a pointer to a singleton ducc_thread_pool, which is always available
+inline ducc_thread_pool *get_master_pool()
   {
-  static thread_pool pool;
+  static auto master_pool = new ducc_thread_pool();
 #if __has_include(<pthread.h>)
   static std::once_flag f;
   call_once(f,
     []{
     pthread_atfork(
-      +[]{ get_pool().shutdown(); },  // prepare
-      +[]{ get_pool().restart(); },   // parent
-      +[]{ get_pool().restart(); }    // child
+      +[]{ get_master_pool()->shutdown(); },  // prepare
+      +[]{ get_master_pool()->restart(); },   // parent
+      +[]{ get_master_pool()->restart(); }    // child
       );
     });
+#endif
+  return master_pool;
+  }
+
+thread_local thread_pool *active_pool = get_master_pool();
+
+#elif DUCC0_LOWLEVEL_THREADING == DUCC0_NO_LOWLEVEL_THREADING
+
+class ducc_pseudo_thread_pool: public thread_pool
+  {
+  public:
+    ducc_pseudo_thread_pool() {}
+
+    // virtual
+    size_t nthreads() const { return 1; }
+
+    // virtual
+    size_t adjust_nthreads(size_t /*nthreads_in*/) const
+      { return 1; }
+    // virtual
+    void submit(std::function<void()> work)
+      { work(); }
+  };
+
+// return a pointer to a singleton ducc_thread_pool, which is always available
+inline ducc_pseudo_thread_pool *get_master_pool()
+  {
+  static auto master_pool = new ducc_pseudo_thread_pool();
+  return master_pool;
+  }
+
+thread_local thread_pool *active_pool = get_master_pool();
+
 #endif
 
-  return pool;
+thread_pool *set_active_pool(thread_pool *new_pool)
+  { return std::exchange(active_pool, new_pool); }
```

Comment:
`active_pool` doesn't necessarily exist (if not one of the above configurations).  Perhaps the `set_active_pool` and `get_active_pool` functions should be moved to inside the `#ifdef`.

Reply:
Good point, I'll thake care of that.

Given how things are now converging, the abstract `threadpool` base class also seems a bit overkill now. As long as every implementation comes with a `threadpool` that presents the required interface, we could get rid of the virtual function overhead, I expect.
That would eliminate the problem above by construction.

Reply:
Sorry, after re-thinking I'm not sure that's correct. If we move the `*_active_pool` functions into the `#ifdef`, `ScopedUseThreadPool` will break.

Reply:
The `set_active_pool` and `get_active_pool` functions themselves need to be defined somewhere, and could be provided by the user if they are using a custom thread pool.   The issue specifically with `active_threadpool` is that it's a global constant, which we can't define in a header without having ODR issues.

You could also move the definition of `active_threadpool` outside the `#ifdef` to solve this - which would then require the user to provide a `get_master_pool()` function.

When playing around with this, this was my `ducc0_threading_config.h` header:
```
#include <mutex>
#include <condition_variable>

namespace ducc0 {
namespace detail_threading {
  
  using Mutex = std::mutex;
  using LockGuard = std::lock_guard<Mutex>;
  using UniqueLock = std::unique_lock<Mutex>;
  using CondVar = std::condition_variable;
  
  class thread_pool;
  
  inline thread_pool** __active_pool() {
    thread_local thread_pool* active_pool = nullptr;
    return &active_pool;
  }
  
  inline thread_pool* set_active_pool(thread_pool* pool) {
    thread_pool* old = *__active_pool();
    *__active_pool() = pool;
    return old;
  }
  
  inline thread_pool* get_active_pool() {
    return *__active_pool();
  }

}  // namespace detail_threading
}  // namespace ducc0
```

Reply:
I think I understand. Does the recent change address this properly?